src/Bicep.Core/Emit/ForSyntaxValidatorVisitor.cs

```diff
@@ -237,20 +283,78 @@ protected override void VisitInternal(SyntaxBase node)
+
+                    if (!isValidResourceCollectionDirectAccessLocation)
+                    {
+                        if (this.variableAccessForInlineCheck != null)
+                        {
+                            IEnumerable<string>? inlineVariableChain = null;
+                            if (this.inlineVariableChainBuilder != null)
+                            {
+                                this.inlineVariableChainBuilder.Add(symbol.Name);
+                                inlineVariableChain = this.inlineVariableChainBuilder.ToImmutable();
+                            }
+                            WriteDirectAccessToCollectionNotSupported(this.variableAccessForInlineCheck, inlineVariableChain);
+                        }
+                        else
+                        {
+                            WriteDirectAccessToCollectionNotSupported(variableOrResourceAccessSyntax);
+                        }
+                    }
+                }
+            } else if ((this.currentVariableDeclarationSyntax == null || this.loopLevel > 0)
+                && variableOrResourceAccessSyntax is VariableAccessSyntax variableAccessSyntax
+                && symbol is VariableSymbol variableSymbol
+                && InlineDependencyVisitor.ShouldInlineVariable(this.semanticModel, variableSymbol.DeclaringVariable, out var outInlineVariableChain))
+            {
+                if (this.inlineVariableVisitLevel == 0)
+                {
+                    this.variableAccessForInlineCheck = variableAccessSyntax;
+                    var chainBuilder = ImmutableArray.CreateBuilder<string>();
+                    chainBuilder.AddRange(symbol.Name);
+                    chainBuilder.AddRange(outInlineVariableChain);
+                    this.inlineVariableChainBuilder = chainBuilder;
+                }
+
+                this.inlineVariableVisitLevel++;
+                VisitVariableDeclarationSyntax(variableSymbol.DeclaringVariable);
+                this.inlineVariableVisitLevel--;
+
+                if (this.inlineVariableVisitLevel == 0)
+                {
+                    this.variableAccessForInlineCheck = null;
+                    this.inlineVariableChainBuilder = null;
                 }
             }
         }
 
-        private static ObjectPropertySyntax? TryGetDependsOnProperty(ObjectSyntax? body) => body?.TryGetPropertyByName("dependsOn");
+        private void WriteDirectAccessToCollectionNotSupported(IPositionable positionable, IEnumerable<string>? accessChain = null) =>
+            this.diagnosticWriter.Write(DiagnosticBuilder.ForPosition(positionable).DirectAccessToCollectionNotSupported(accessChain));
+
+        private static ObjectPropertySyntax? TryGetDependsOnProperty(ObjectSyntax? body) => body?.TryGetPropertyByName(LanguageConstants.ResourceDependsOnPropertyName);
+        private static ObjectPropertySyntax? TryGetPropertiesProperty(ObjectSyntax? body, bool isModule) =>
```

Comment:
I'd suggest eliminating the ternary operator by spliting the method into `TryGetResourcePropertiesProperty(ObjectSyntax?)` and `TryGetModuleParamsProperty(ObjectSyntax?)`.

Reply:
https://github.com/Azure/bicep/pull/10939/commits/06bef2d9666f2c28df579a89fd8cdcb602a8ba0e